Objects/setobject.c

```diff
@@ -2301,12 +2490,23 @@ static PyObject *
 frozenset___contains___impl(PySetObject *so, PyObject *key)
 /*[clinic end generated code: output=2301ed91bc3a6dd5 input=2f04922a98d8bab7]*/
 {
-    long result;
-
-    result = set_contains_lock_held(so, key);
-    if (result < 0)
+    Py_hash_t hash = _PyObject_HashFast(key);
+    if (hash == -1) {
+        if (!PySet_Check(key) || !PyErr_ExceptionMatches(PyExc_TypeError)) {
+            set_unhashable_type(key);
+            return NULL;
+        }
+        PyErr_Clear();
+        Py_BEGIN_CRITICAL_SECTION(key);
+        hash = frozenset_hash_impl(key);
+        Py_END_CRITICAL_SECTION();
```

Comment:
Note: this code looks a bit odd: why add a critical section around a call to `frozenset_hash_impl`? The reason is that `key` can be both a `set` and `frozenset`. The `frozenset_hash_impl` handles the `set`, but to make it thread safe we add a critical section. The code is the same in current main (but moved around a bit)

In python testing for membership in a `set` with an object of type `set` is allowed, even though a `set` is not hashable. E.g. `set() in {frozenset(), frozenset([1,2])}` returns `True`. The same does not hold for other container types, e.g. `set() in {frozenset(): 2}` returns a `TypeError`.

@nascheme Maybe add a comment to clarify?

Reply:
Comment added.